include/clap/ext/draft/scratch-memory.h

```diff
@@ -0,0 +1,90 @@
+#pragma once
+
+#include "../../plugin.h"
+
+// This extension lets the plugin request "scratch" memory
+// from the host. Scratch memory can be accessed during the
+// `process()` callback, but is not persistent between callbacks.
+//
+// The motivation for this extension is to allow the plugin host
+// to "share" a single scratch buffer across multiple plugin
+// instances.
+//
+// For example, imagine the host needs to process three plugins
+// in sequence, and each plugin requires 10K of scratch memory.
+// If each plugin pre-allocates its own scratch memory, then 30K
+// of memory is being allocated in total. However, if each plugin
+// requests 10K of scratch memory from the host, then the host can
+// allocate a single 10K scratch buffer, and make it available to all
+// three plugins.
+//
+// On memory-constrained platforms, this optimization may allow for
+// more plugins to be used simultaneously. On platforms with lots
+// of memory, this optimization may improve CPU cache usage.
+
+static CLAP_CONSTEXPR const char CLAP_EXT_SCRATCH_MEMORY[] = "clap.scratch-memory/1";
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct clap_host_scratch_memory {
+   // Asks the host for certain amount of scratch memory.
+   // If the host is unable to provide the memory, it should
+   // return "false".
+   //
+   // The plugin may call this method multiple times (for
+   // example, gradually decreasing the amount of scratch
+   // being asked for until the host returns true), however,
+   // the plugin should avoid calling this method un-neccesarily
+   // since the host implementation may be relatively expensive.
+   // If the plugin calls `reserve()` multiple times, then the
+   // final call determines the actual amount of scratch memory
+   // that will be available to the plugin. If the final call
+   // returns false then no scratch memory will be provided,
+   // regardless of any previous calls to `reserve()`.
+   //
+   // Note that any memory the host allocates to satisfy
+   // the requested scratch size can be de-allocated by the
+   // host when the plugin is de-activated.
+   //
+   // In the context of plugins and hosts that implement
+   // the "thread-pool" extension, scratch memory is assumed
+   // to be "thread-local". The plugin should request the maximum
+   // amount of scratch memory that it will need on a single
+   // thread. Accordingly, the host must ensure that each
+   // thread can independently provide the requested amount
+   // of scratch memory.
```

Comment:
thought: If I understand correctly, I think this could be challenging for a host to implement efficiently, as I'm not sure the host necessarily knows during the call to `reserve` how many concurrent jobs a plugin is going to request during `process`, and therefore how much memory needs to be allocated if it wants to be able to run every job in parallel. Consequently, the host would be forced to allocate the worst-case amount of requested memory for every thread pool thread it creates, which might be much _more_ memory than is actually required.

As an extreme example, if the plugin knows it is _never_ going to create thread pool jobs, then all of the per-thread scratch buffers allocated by the host for that plugin are wasted.

I wonder if it would be worth augmenting `reserve` to have a `uint32_t concurrent_buffers` argument, which;
- If `0`, means that all tasks receive the same buffer from `access` and it is up to the plugin to deal with that as it wishes (perhaps even a feature)
- If `n > 0` means that task indices up to `n - 1` will successfully receive a buffer pointer upon calling `access`, and task indices `n` and upwards may receive NULL?

Reply:
Adding a `concurrent_buffers` argument seems reasonable to me... @abique do you have any thoughts on this?

It would be nice if we could have that argument default to zero so plugins that don't implement `thread-pool` don't need to worry about it, but I know that's not possible for a C API.

Reply:
You never know as well on which thread of your thread pool you're going to call `process()`.
All your thread pool threads needs to have a thread local scratch memory that is `>=all_plugin_instance_max_scratch_size`.

Reply:
I'm not sure about `concurrent_buffers`.

At least it should be no more than a hint, and shouldn't change the behavior. Because we already have a clear semantic, where the plugin can fetch the scratch buffer before calling into the thread pool.
So for me, `0` means no hint.

The only use case for it would be that you are on a crazy modern CPU, like 192 arm machine, and you don't want to allocate 192 times the max scratch size, so you have some kind of pool of scratch buffers that you can pick and release into in a lock-free way 🤔 

I'm not sure, I think that a scratch size of up to 64KB is acceptable, and I'd simply allocate one for each thread and have a simple a robust thread local access, instead of a dodgy lock-free pool of scratch buffers.

If you have a 192 cores CPU, you probably have 32GB of ram if not more, so 192*64KB should be acceptable as well.

Reply:
Still, the `concurrent_buffers` doesn't convince me, because the goal is to share the buffer with other plugins.

If you have N plugins that says max M tasks, the num of scratch buffers that you need is `min(M*N, thread_pool_size)`.

The scratch buffer is the most beneficial when N grows, because it shares the most memory.
While `concurrent_buffers` is the most beneficial when N is small.

I don't see a way to take advantage of `concurrent_buffers` that is worth the pain.

Reply:
> You never know as well on which thread of your thread pool you're going to call process().
> All your thread pool threads needs to have a thread local scratch memory that is >=all_plugin_instance_max_scratch_size.

I'm not sure that's necessarily true: it depends on how the host implements the thread pool. There are valid use cases that means that the host *does* know which threads are going to be used by which plugins.

Reply:
> At least it should be no more than a hint, and shouldn't change the behavior

That would defeat the point: the motivation for the suggested change is to avoid the host being forced to allocate `max(any_plugin_scratch)` for _every_ thread pool thread it creates. The change allows the host to only allocate as many buffers as are actually going to be needed.

> and I'd simply allocate one for each thread and have a simple a robust thread local access

As a host you could still choose to do that - the `concurrent_buffers` request doesn't prevent the host from providing _more_ buffers than the plugin requests; in the way I worded it, tasks from index `n` upwards _may_ receive `NULL` (but _may_ still receive a buffer, if the host just chooses to allocate the max scratch for every thread in the system).

Reply:
I think the key thing comes back to what I mentioned above RE the host needing to allocate memory for the threads which is not actually needed. Imagine the following flow where there is one homogeneous thread pool and the host has no concept of the number of concurrent scratch buffers are needed by a plugin.

1. Decide that you are going to allow for 8 audio worker threads. Create a thread pool with 8 threads.
2. Activate plugin A which requests `X` bytes of scratch memory. Allocate `(1+8)X` bytes of scratch memory; 1 for the audio thread, and one for each worker thread, and distribute scratch pointers as appropriate.
3. Start processing. Plugin A uses the 8 worker threads.
4. Some time later, activate plugin B which requests `Y > X` bytes of scratch memory.
5. The host must now reallocate the scratch buffer allocation to be `(1+8)Y` bytes, do the pointer juggling with the worker threads, and deallocate the old buffers.
6. Continue processing. Plugin A continues to use the 8 worker threads as before. Plugin plugin B, as it happens, doesn't need to use any worker threads.

Result: The host has allocated `(1+8)Y` bytes, whereas it only needed to allocate `Y + 8X`. If `Y >> X`, this becomes quite wasteful quite quickly.

If plugin B had been able to signal `concurrent_buffers == 0`:
* the total memory usage would be lower, as the host needs less memory allocated in step 5, and;
* if `Y <= (1 + 8)X`, the host can avoid doing any of the work in step 5 at all if it treats the entire scratch buffer initially allocated as one large arena.

---

If a host isn't interested in supporting any of these optimisations, they can just ignore the non-zero value of `concurrent_buffers` and allocate away merrily, so I don't see a real downside from the host's perspective?

Reply:
I agree that there doesn't seem to be a down-side to informing the host about the number of `concurrent_buffers` the plugin might need, aside from the API being slightly more complex.

Even if the host may not have a way to use that information, I think it makes sense for the plugin to be as specific as possible about what it's asking for.

Reply:
@p--b thanks, I now get the idea behind your scenario and I see how it could be happening.

@jatinchowdhury18 then go for it! 👍

Reply:
@jatinchowdhury18 maybe you should add a comment that if the plugin asks for N concurrent buffers, but actually uses N+1, then what happens, crash? nullptr?

@p--b would `bool is_concurrent` suffice?

Reply:
> It would be nice if we could have that argument default to zero so plugins that don't implement `thread-pool` don't need to worry about it, but I know that's not possible for a C API.

It is possible...by making it a separate extension. ;)

Reply:
Sorry I missed your comment! Making it a `bool` would help signal to the host in the case where the plugin isn't going to use the threadpool at all, but doesn't unlock as many optimisations for the host if the plugin is using threadpool tasks: if the plugin knows it's only ever going to fire off 4 threadpool tasks, then it's potentially wasteful for the host to allocate memory into all N threadpool threads where it only needs to allocate into 4. Really I guess the ideal would be some way for the plugin to indicate how many parallel threadpool tasks it wants to execute _a priori_ as part of the threadpool extension, but without that, at least allowing it to be indicated here allows the host not to have to overallocate. That said, if it was `bool` or nothing, then `bool` is definitely better :)